core/aws_ddk_core/stages/sns_lambda.py

```diff
@@ -0,0 +1,202 @@
+            Amount of errored function invocations before triggering CW alarm. Defaults to `5`
+        lambda_function_errors_alarm_evaluation_periods: Optional[int]
+            The number of periods over which data is compared to the specified threshold. Defaults to `1`
+        function_props : Any
+            Additional function properties. For complete list of properties refer to CDK Documentation -
+            Lambda Function: https://docs.aws.amazon.com/cdk/api/v2/python/aws_cdk.aws_lambda/Function.html
+        topic_props : Any
+            Additional topic properties. For complete list of properties refer to CDK Documentation -
+            SNS Topic: https://docs.aws.amazon.com/cdk/api/v2/python/aws_cdk.aws_sns/Topic.html
+        alarms_enabled: Optional[bool]
+            Enable/Disable all alarms in the stage.
+            Default - True
+        """
+        super().__init__(scope, id, alarms_enabled=alarms_enabled)
+
+        self._event_source: str = f"{id}-event-source"
+        self._event_detail_type: str = f"{id}-event-type"
+
+        if lambda_function:
+            self._function = lambda_function
+        elif code and handler:
+            self._function = LambdaFactory.function(  # type: ignore
+                self,
+                id=f"{id}-function",
+                environment_id=environment_id,
+                code=code,
+                handler=handler,
+                runtime=runtime,
+                role=role,
+                memory_size=memory_size,
+                timeout=timeout,
+                environment={
+                    "EVENT_SOURCE": self._event_source,
+                    "EVENT_DETAIL_TYPE": self._event_detail_type,
+                },
+                layers=layers,
+                **function_props,
+            )
+        else:
+            raise ValueError("'code' and 'handler' or 'lambda_function' must be set to instantiate this stage")
+
+        # Enable the function to publish events to the default EventBus
+        self._function.add_to_role_policy(
+            PolicyStatement(
+                effect=Effect.ALLOW,
+                actions=[
+                    "events:PutEvents",
+                ],
+                resources=["*"],
+            )
+        )
+
+        self._topic = sns_topic or SNSFactory.topic(  # type: ignore
+            self,
+            id=f"{id}-topic",
+            environment_id=environment_id,
+            topic_name=topic_name,
+            **topic_props,
+        )
+
+        self._function.add_event_source(
+            SnsEventSource(
+                topic=self._topic,
+                dead_letter_queue=SQSFactory.queue(
+                    self,
+                    id=f"{id}-dlq",
+                    environment_id=environment_id,
+                )
+                if dead_letter_queue_enabled
+                else None,
+                filter_policy=filter_policy,
+            )
+        )
```

Comment:
Should we have an SQS queue in between the SNS topic and the Lambda to make the lambda more efficient? Also, have you tested how the redrive  logic works with just an SNS queue? Is there an easy way to republish a message from the DLQ into the SNS topic?